File: C_Planner/common/cost/cost_function.cpp
#include "cost_function.h"
#include <limits>

CostFunction::CostFunction(const std::string& cost_type) {
    w_T = 1.0;
    w_V = 0.1;
    w_A = 10.0;
    w_J = 10.0;
    w_D = 100.0;
    w_LC = 10.0;
    w_dist = 0.1;
    max_speed = 14.0;
}

double CostFunction::cost_terminal_time(double terminal_time) {
    return w_T * terminal_time;
}

double CostFunction::cost_velocity_offset(const std::vector<double>& vels, double v_target) {
    double cost = 0.0;
    for (double vel : vels) {
        double diff = vel - v_target;
        cost += diff * diff;
    }
    return w_V * cost;
}

double CostFunction::cost_acceleration(const std::vector<double>& accels) {
    double cost = 0.0;
    for (double accel : accels) {
        cost += accel * accel;
    }
    return w_A * cost;
}

double CostFunction::cost_jerk(const std::vector<double>& jerks) {
    double cost = 0.0;
    for (double jerk : jerks) {
        cost += jerk * jerk;
    }
    return w_J * cost;
}

double CostFunction::cost_lane_center_offset(const std::vector<double>& offsets) {
    double cost = 0.0;
    for (double offset : offsets) {
        cost += offset * offset;
    }
    return w_LC * cost;
}

double CostFunction::cost_dist_obstacle(const double* obstacles_array,
                                        const int* num_vertices_array,
                                        int num_time_steps,
                                        int num_obstacles,
                                        int max_vertices,
                                        const FrenetTrajectory& traj,
                                        int time_step_now) {
    if (obstacles_array == nullptr || num_vertices_array == nullptr ||
        num_time_steps <= 0 || num_obstacles <= 0 || max_vertices <= 0 ||
        traj.x.empty() || traj.y.empty()) {
        return 0.0;
    }

    const size_t num_traj_points = std::min(traj.x.size(), traj.y.size());
    double exp_sum = 0.0;

    for (size_t i = 0; i < num_traj_points; ++i) {
        const int t_idx = time_step_now + i;
        if (t_idx < 0 || t_idx >= num_time_steps) {
            break;
        }

        const double traj_x = traj.x[i];
        const double traj_y = traj.y[i];
        double min_dist = std::numeric_limits<double>::infinity();

        for (int obs_idx = 0; obs_idx < num_obstacles; ++obs_idx) {
            const int num_verts = num_vertices_array[t_idx * num_obstacles + obs_idx];
            const int valid_num_verts = std::min(num_verts, max_vertices);
            if (valid_num_verts <= 0) {
                continue;
            }

            double center_x = 0.0;
            double center_y = 0.0;
            for (int v = 0; v < valid_num_verts; ++v) {
                const int array_idx = (t_idx * num_obstacles + obs_idx) * max_vertices * 2 + v * 2;
                center_x += obstacles_array[array_idx];
                center_y += obstacles_array[array_idx + 1];
            }
            center_x /= static_cast<double>(valid_num_verts);
            center_y /= static_cast<double>(valid_num_verts);

            const double dist = std::hypot(traj_x - center_x, traj_y - center_y);
            if (dist < min_dist) {
                min_dist = dist;
            }
        }

        if (std::isfinite(min_dist)) {
            exp_sum += std::exp(-min_dist * w_dist);
        }
    }

    return w_D * exp_sum;
}

double CostFunction::cost_singleTrajectory(const FrenetTrajectory& traj,
                                           double target_speed,
                                           const double* obstacles_array,
                                           const int* num_vertices_array,
                                           int num_time_steps,
                                           int num_obstacles,
                                           int max_vertices,
                                           int time_step_now) {
    if (traj.t.empty()) {
        return std::numeric_limits<double>::infinity();
    }

    double cost_time = cost_terminal_time(15.0 - 0.1 * static_cast<double>(traj.t.size()));
    double cost_obstacle = cost_dist_obstacle(obstacles_array, num_vertices_array, num_time_steps, num_obstacles, max_vertices, traj, time_step_now);
    double cost_speed = cost_velocity_offset(traj.s_d, max_speed);
    double cost_accel = cost_acceleration(traj.s_dd) + cost_acceleration(traj.d_dd);
    double cost_jerk_val = cost_jerk(traj.s_ddd) + cost_jerk(traj.d_ddd);
    double cost_offset = cost_lane_center_offset(traj.d);

    return (cost_time + cost_obstacle + cost_speed + cost_accel + cost_jerk_val + cost_offset) /
           static_cast<double>(traj.t.size());
}

void CostFunction::calc_cost(std::vector<FrenetTrajectory>& fplist,
                             double target_speed,
                             const double* obstacles_array,
                             const int* num_vertices_array,
                             int num_time_steps,
                             int num_obstacles,
                             int max_vertices,
                             int time_step_now) {
    for (auto& traj : fplist) {
        traj.cost_final = cost_singleTrajectory(
            traj,
            target_speed,
            obstacles_array,
            num_vertices_array,
            num_time_steps,
            num_obstacles,
            max_vertices,
            time_step_now
        );
    }
}

double CostFunction::cost_total(const FrenetTrajectory& traj, double target_speed) {
    (void)target_speed;

    if (traj.t.empty()) {
        return std::numeric_limits<double>::infinity();
    }

    double cost_time = cost_terminal_time(15.0 - 0.1 * static_cast<double>(traj.t.size()));
    double cost_obstacle = 0.0;
    double cost_speed = cost_velocity_offset(traj.s_d, max_speed);

    double cost_accel = cost_acceleration(traj.s_dd) + cost_acceleration(traj.d_dd);
    double cost_jerk_val = cost_jerk(traj.s_ddd) + cost_jerk(traj.d_ddd);
    double cost_offset = cost_lane_center_offset(traj.d);

    double total_cost = (cost_time + cost_obstacle + cost_speed + cost_accel + cost_jerk_val + cost_offset) /
                        static_cast<double>(traj.t.size());
    
    return total_cost;
}

File: fiss_plus_planner/planners/common/cost/cost_function.py
import numpy as np
from fiss_plus_planner.planners.common.scenario.frenet import FrenetTrajectory

class CostFunction:
    def __init__(self, cost_type: str):
        if cost_type is "WX1":
            self.w_T = 10.0
            self.w_V = 0.1
            self.w_A = 1.0
            self.w_J = 1.0
            self.w_D = 100.0
            self.w_LC = 10.0
            self.w_dist = 1.0
            self.max_speed = 14.0  # for normalization, can be adjusted based on

    
    def cost_time(self) -> float:
        pass
    
    def cost_terminal_time(self, terminal_time: float) -> float:
        return self.w_T * terminal_time
    
    def _compute_polygon_center(self, vertices: np.ndarray, num_verts: int) -> np.ndarray:
        if num_verts <= 0:
            return np.array([np.inf, np.inf])
        return vertices[:num_verts].mean(axis=0)

    def is_obstacle_front(self, ego_pose: np.ndarray, obs_center: np.ndarray) -> bool:
        # returns whether the obstacle is in front of the ego vehicle
        # based on the dot product between the relative vector between ego and obstacle and the ego direction vector
        # if > 0, then the obstacle is in front, otherwise it's behind
        
        ego_direction = np.array([np.cos(ego_pose[2]), np.sin(ego_pose[2])])
        relative_vector = obs_center - ego_pose[:2]
        dot_product = np.dot(relative_vector, ego_direction)
        # print(dot_product)
        return dot_product > 0
    
    def cost_dist_obstacle(
        self, 
        obstacles_array: np.ndarray, 
        obstacles_num_vertices: np.ndarray,
        traj: FrenetTrajectory, 
        time_step_now: int = 0
    ) -> float:
        num_traj_points = len(traj.x)
        num_time_steps = obstacles_array.shape[0]
        num_obstacles = obstacles_array.shape[1]
        
        if num_obstacles == 0 or num_traj_points == 0:
            return 0.0
        
        min_dists = []
        
        for i in range(num_traj_points):
            t_idx = time_step_now + i
            # t_idx = time_step_now
            
            if t_idx >= num_time_steps:
                break
                
            traj_x = traj.x[i]
            traj_y = traj.y[i]
            
            min_dist = np.inf
            
            for obs_idx in range(num_obstacles):
                num_verts = obstacles_num_vertices[t_idx, obs_idx]

                if num_verts <= 0:
                    continue
                
                vertices = obstacles_array[t_idx, obs_idx]
                center = self._compute_polygon_center(vertices, num_verts)
                
                dist = np.sqrt((traj_x - center[0])**2 + (traj_y - center[1])**2)
                
                if dist < min_dist:
                    min_dist = dist
            
            if min_dist < np.inf:
                min_dists.append(min_dist)
        
        if len(min_dists) == 0:
            return 0.0
        
        Xis = np.exp(-np.array(min_dists)*self.w_dist)
        return self.w_D * np.sum(Xis)
    
    def cost_velocity_offset(self, vels: list, v_target: float) -> float:
        return self.w_V * sum(np.power(np.subtract(vels, v_target), 2))
    
    def cost_acceleration(self, accels: list) -> float:
        return self.w_A * sum(np.power(accels, 2))
            
    def cost_jerk(self, jerks: list) -> float:
        return self.w_J * sum(np.power(jerks, 2))
    
    def cost_lane_center_offset(self, offsets: list) -> float:
        return self.w_LC * sum(np.power(offsets, 2))
    
    def cost_total(self, traj: FrenetTrajectory, target_speed: float) -> float:
        cost_time = self.cost_terminal_time(15.0 - 0.1*len(traj.t))
        cost_obstacle = 0.0 # self.cost_dist_obstacle()
        cost_speed = self.cost_velocity_offset(np.abs(traj.s_d), self.max_speed)
        cost_accel = self.cost_acceleration(traj.s_dd) + self.cost_acceleration(traj.d_dd)
        cost_jerk = self.cost_jerk(traj.s_ddd) + self.cost_jerk(traj.d_ddd)
        cost_offset = self.cost_lane_center_offset(traj.d)
        # return cost_speed + cost_accel + cost_jerk + cost_offset
        cost_total = (cost_time + cost_obstacle + cost_speed + cost_accel + cost_jerk + cost_offset)/len(traj.t)
        return cost_total
    
    def cost_singleTrajectory(
        self, 
        traj: FrenetTrajectory, 
        target_speed: float, 
        obstacles_array: np.ndarray,
        obstacles_num_vertices: np.ndarray,
        time_step_now: int
    ) -> float:
        cost_time = self.cost_terminal_time(15.0 - 0.1*len(traj.t)) 
        cost_obstacle = 0.0 #self.cost_dist_obstacle(obstacles_array, obstacles_num_vertices, traj, time_step_now)
        cost_speed = self.cost_velocity_offset(np.abs(traj.s_d), self.max_speed)
        cost_accel = self.cost_acceleration(traj.s_dd) + self.cost_acceleration(traj.d_dd)
        cost_jerk = self.cost_jerk(traj.s_ddd) + self.cost_jerk(traj.d_ddd)
        cost_offset = self.cost_lane_center_offset(traj.d)
        cost_total = (cost_time + cost_obstacle + cost_speed + cost_accel + cost_jerk + cost_offset)/len(traj.t)
        return cost_total
    
    def final_trajectory_cost(self,
                              traj: FrenetTrajectory,
                              obstacles_array: np.ndarray,
                              obstacles_num_vertices: np.ndarray) -> float:
        """
        calculate the cost of the final executed trajectory
        """
        
        cost_time = self.cost_terminal_time(0.1*len(traj.t))  # assuming each time step is 0.1s, and the total time is 10s
        cost_obstacle = self.cost_dist_obstacle(obstacles_array, obstacles_num_vertices, traj)
        cost_speed = self.cost_velocity_offset(np.abs(traj.v), self.max_speed)
        cost_accel = self.cost_acceleration(traj.s_dd) + self.cost_acceleration(traj.d_dd)
        cost_jerk = self.cost_jerk(traj.s_ddd) + self.cost_jerk(traj.d_ddd)
        cost_offset = self.cost_lane_center_offset(traj.d)
        cost_total = ( cost_obstacle + cost_speed + cost_accel + cost_jerk + cost_offset)/len(traj.t) + cost_time
        # print("velocity:", traj.v)
        print("Cost Distance To Obstacles:", cost_obstacle)
        print("Cost Time:", cost_time)
        print("Cost Speed:", cost_speed)
        print("Cost Acceleration:", cost_accel)
        print("Cost Jerk:", cost_jerk)
        print("Cost Lane Center Offset:", cost_offset)
        print("len(traj.t):", len(traj.t))
        print("Total Cost:", cost_total)
        return cost_total
    
    def calc_cost(self, fplist:list, target_speed: float, obstacles_array: np.ndarray,
        obstacles_num_vertices: np.ndarray, time_step_now: int) -> list:
        for traj in fplist:
            traj.cost_final = self.cost_singleTrajectory(traj, target_speed, obstacles_array, obstacles_num_vertices, time_step_now)
        return fplist

